### DISCUSSION TITLE
I am willing to pay for feature implementation, how to go about it?

### DISCUSSION THREAD

**Discussion by SilencedPerson** [godotengine/godot-proposals#4298] [Engine Core]
Let's say I want small feature implemented, don't know how to go about it myself but I am willing to pay someone to do it for me.
How do I obtain blessing from the devs?

**Comment by Mickeon** [comment]
Well, first and foremost, you should probably gather blessings from the userbase by proposing it, then the developers should approve it, because Godot is a community effort of volunteers, at heart.

**Reply by SilencedPerson** [reply]
It is for feature frequently requested by the community. I am working with assumption that I already have a blessing of the community.

**Reply by Calinou** [reply]
@SilencedPerson Which feature do you have in mind? Could you link to the proposal here?

What Godot's paid contractors work on is set in advance, so you cannot steer them towards working on specific features you'd like. [Paid contractors are paid to do what they already do well](https://godotengine.org/governance), so it would be less productive if contractors were tasked to do stuff they're unfamiliar with.

If you'd like to pay a third-party developer to work on a feature instead, you don't need to seek approval here. However, we can't be held responsible if the person in question ends up not doing the work (or botches it). We also can't make guarantees that any PR will be merged.

**Reply by SilencedPerson** [reply]
https://github.com/godotengine/godot-proposals/issues/4282
https://github.com/godotengine/godot-proposals/issues/2710

Here is quick google search where people talk about the problem and current way of achieving is very convoluted and requires shaders (even if very simple).
https://www.reddit.com/r/godot/comments/o8pp6m/easy_way_to_mask_a_sprite_in_2d_differently_from/

I teach teens how to use Godot and we keep running into this issue where the current masking blending mode of Light2D does exactly the opposite thing that they expect it to do. The feature i am looking to have implemented is just inverting the current behavior of the masking blending mode.

![image](https://user-images.githubusercontent.com/82024211/160413628-c8d088aa-5621-4591-ab9e-e899e577435c.png)

Making masks that are not simple shapes is kind of a non-trivial matter with the way things are done currently.

**Reply by Calinou** [reply]
CanvasItem clipping should be fixed first before we can look into another implementation.

**Reply by SilencedPerson** [reply]
This has nothing do to with clipping. Clipping is this and that is not what feature is about.
![tIXUpAmlTJ](https://user-images.githubusercontent.com/82024211/160418458-f9fb1015-8085-4c34-94a9-d8e427fb8902.gif)

**Comment by theraot** [comment]
Anybody can make a fork, implement a feature or fix a bug, and make a pull request. They don't need anybody's approval or permission. Furthermore, if there is already a proposal or an issue, then it will go in the usual review process, and it might get merged.

This means that you don't need anybody approval or blessing to go hire somebody to work on a feature. Heck, it is your money. So you hire a developer, you know, the same way you would hire a developer to work for you, because that is what you would be doing.

If you are saying you want Godot to hire somebody to work on your feature… I believe that is a no. You can throw money at Godot (e.g. donations), but that shall not give you power over Godot. It does not mean they will not work on it, but it does not mean they will work on it either.

If you prefer to learn how to contribute instead, start with the documentation there is a section on Contributing. Also you can find help via the community sites (go to the official site and click on Community). *And by the way, this is not the Q&A site, this space is for discussing proposals.*

*By the way, I suspect you don't really know how hard it is to implement the feature you want. You say "small", but you don't know how to implement it… You say it is "frequently requested", and has nobody step up to implement it? I'm warning you of unexpected challenges.*

**Reply by SilencedPerson** [reply]
Currently, people are being dismissed with "Light2D already does this".
It does, but in a flawed way.

**Reply by Mickeon** [reply]
By all means then, support the existing proposals or create a new one. If it's referring to [this](https://github.com/godotengine/godot-proposals/issues/4282), the proposal is gathering a lot of positive feedback and it's worth at the very least noting it for the future.

**Reply by Zireael07** [reply]
@Mickeon this is exactly what he is talking about

**Reply by mrjustaguy** [reply]
as Calinou stated above, you may seek a third party dev to implement a feature, but Godot might not merge it for any number of issues, be it a bad implementation, bloat that almost no one cares about (in which case it's a good candidate for a niche plugin often times) or lack of consensus.

The best idea is to find if a Proposal for the feature already exists, or if it doesn't, make one and see the feedback before hiring someone, unless you're fine with it being in a fork of Godot in case it gets rejected..

also, theroat, one could understand the principle behind how the feature works and not be able to implement due to lacking the knowledge with the required programming language, and frequently requested and nobody implementing could also be because they too don't have the knowledge to code it in a specific language, and/or they don't have the time for whatever reason.. I've seen plenty of both cases, not as uncommon as you might think.

**Reply by Calinou** [reply]
I generally recommend experimenting with new features by implementing them in an add-on. This lets you see if there's enough people that would actually use the system as implemented by your add-on, and also lets you catch potential limitations of the proposed system.

This can also speed up the implementation of the feature in core Godot by providing a point of reference.

**Reply by Xrayez** [reply]
> This can also speed up the implementation of the feature in core Godot by providing a point of reference.

That's why I constantly ask Godot team to promote Goost and other projects via Godot's documentation, see #3864.

**Reply by Zireael07** [reply]
BTW Goost isn't the only one, I also found a fork named 'Goblin Engine' :)

**Reply by theraot** [reply]
Speaking of forks, I'll remind you of Godex: https://github.com/GodotECS/godex not of particular relevance for the topic of this thread. Edit: Oh, and the DragonBones one: https://ozzadar.itch.io/godot-dragonbones-3-3-3 - Ok, Ok, those are modules, I get it, but how do you develop modules? Yes, you fork. Until GDExtension, I think.

**Reply by Calinou** [reply]
> Ok, Ok, those are modules, I get it, but how do you develop modules? Yes, you fork.

Developing a statically compiled C++ module doesn't require forking the engine. The module can be placed in a self-contained repository, cloned by the user and then built with the engine :slightly_smiling_face:

**Reply by theraot** [reply]
@Calinou I think I can make an addon based on the approach from reddit: https://www.reddit.com/r/godot/comments/o8pp6m/easy_way_to_mask_a_sprite_in_2d_differently_from/ But I don't think a proper solution would have similarity with it, as this is rendering the background on top instead of not rendering in the first place.

I could render the mask to another viewport, which is an approach I've explained over StackOverflow a while back: https://stackoverflow.com/a/69916438/402022 but I don't think I can make that addon… Or well, I could, but I could not have the addon simply add the custom shader as next pass because a pass cannot hide prior passes rendered. Similarly, it cannot be inserted as first pass because the others would simply draw on top of it. So I'd need to modify the material, and that is not trivial.

If I had access to the way light is computed from an addon, I would do it that way instead. But I can't.

Now, here is an idea: modulate mask. Again, that can't be an addon, but if the light approach is not good, this could be another way to go about it.

I really don't know where to start when it comes to shaders in Godot source. And sadly, I don't think addons are going to help decide what is the best approach this time.

---

Edit: Oh, I didn't see you replied to my prior comment. I suppose you are right, you don't necessarily have to fork to make a module. I would fork, though, I find it easier.

**Reply by Xrayez** [reply]
> Oh, I didn't see you replied to my prior comment. I suppose you are right, you don't necessarily have to fork to make a module. I would fork, though, I find it easier.

If you only need to integrate one feature and do this for development purposes, it's easier. But if you plan to use your feature in the long term, you'd have to think about maintaining your fork by synchronizing it with upstream Godot repository, and this is *not* an easy task, otherwise you'd be missing out on critical bug fixes from upstream. Not to mention the fact that you'd be limiting potential developers from using your feature if you'd like to publish/share it with others.

Some modules, like [Godex](https://github.com/GodotECS/godex), do require custom changes to the original Godot to function properly. But most can be done in a self-contained way, like [Goost](https://github.com/goostengine/goost). And yeah, those are definitely not forks.

**Comment by Xrayez** [comment]
I'm considering contractual work.

My areas of expertise and interest are (mostly 2D development):
- core engine development
    - custom data structures, the most prominent example is a [`Graph`](https://github.com/godotengine/godot-proposals/issues/3848#issuecomment-1039496859)
- [computational geometry](https://github.com/goostengine/goost/pull/39)
- [procedural generation](https://github.com/Xrayez/godot-anl)
- [physics](https://github.com/Xrayez/box2d-lite-gdscript)
- image processing
- artificial intelligence (AI)

Some features can be implemented directly in [Goost](https://goostengine.github.io/) if they match the vision behind the project: goostengine/goost#189. But I'm fine developing features in other ways outside of the project: GDScript plugins, GDNative/GDExtension, C++ modules (I have much greater experience with C++ modules though). 🙂

In the past, I used to contribute to Godot for the sake of learning and experience (and bug fixing of course, whenever I stumbled upon various limitations while working on my game project). Nowadays, I find myself capable of implementing a variety of features, but figured that I can no longer allow myself to work on random things for the sake of it.

I may be slow to implement something, but eventually I get things done (again, depending on motivation and previous experience). A lot of things do require learning about the domain, which takes time.

That said, you don't need a blessing from the core Godot developers and everything doesn't have to revolve around Godot community exclusively! Me and other interested contributors have already started to build a community around [Goost](https://github.com/goostengine/goost) that aims to complement Godot, mostly independent from Godot decisions, which is nice because Godot's development is no longer limited to one decision-making bottleneck.

**Note**: *Even though I do have over 5 years contribution and maintenance experience with Godot, I'm not affiliated with the official Godot core developers, please don't interpret my words as official Godot announcement, this may lead to negative consequences for me personally, thanks for understanding.*

**Reply by SilencedPerson** [reply]
Thank you, for now I dropped the idea and went back to using game maker in the classes.

**Reply by Xrayez** [reply]
I've been playing with GameMaker at teenage btw. I wonder how much it improved by now. 😃

The way I see it, Godot is certainly good for prototyping purposes and if you're willing to get your hands dirty (that said, maintain your own fork), you can even make a commercially viable game with it in the long term. Hopefully Godot will remain this ability at the very least. Other than that, I still think that Godot misses vital features, if we talk about a certain level Godot aspires for on the surface.

**Reply by SilencedPerson** [reply]
I feel like the workflow got worse in the GMS2 which is why I decided to give Godot a try. Been teaching Godot for 2 years now and while the engine much improved, there are some underlying issues that persist, The changing of icons on exported projects is pain, custom drawing is nearly useless for my purposes in the class, it is simply not flexible enough. For 3D in the advanced classes, I still teach Unity as that is popular demand, that may change however with Unreal Engine 5 being a thing which would finally force me to learn C++.

Godot has been surprisingly good for Open Source software so I plan on giving it a second try when 4.1 or 4.2 drop, it's been very popular with the students but I also feel like some things are way too low level and it definitely did scare off some kids.
